docs/decisions/0002-feature-based-application-organization.rst

```diff
@@ -0,0 +1,157 @@
+
+Secondary principles
+====================
+
+**A. Separate business logic from components**
+
+In order to isolate our view layer (react) from business logic inherent in our global state, APIs, and side effect management, we want to adopt the "ducks" organization (see references).  This involves isolating business logic into a
+sub directory of a feature.  See "B" below for an example.
+
+**B. Files in a feature directory are named by type of code**
+
+Since directory names describe features, file names describe types of code.
+
+::
+
+  /profile
+    /index.js // public interface
+    /ProfilePage.jsx // view
+    /ProfilePhotoUploader.jsx // supporting view
+    /ducks // business logic
+      /actions.js
+      /reducers.js
+      /selectors.js
+
+Within the ducks directory, if multiple files of a type are necessary, first consider whether there is really a sub-feature here that should be defined.  If no discernable sub-feature is emerging, then turn the file (e.g., reducers.js) into a directory (reducers/)
+and put multiple files in it, named by what the reducer does (reducers/AccountReducer.js).  The directory can then have an index.js file which documents the reducers that can be imported.  Consider again whether there's really a sub-feature here.
+
+Multiple components and reducers:
+
+::
+
+  /profile
+    /index.js // public interface
+    /ProfilePage.jsx
+    /ProfilePhotoUploader.jsx
+    /Avatar.jsx
+    /ducks
+      /actions.js
+      /reducers
+        /index.js
+        /FormReducers.js
+        /ProfilePhotoReducers.js
+      /selectors.js
+
+As a sub component:
+
+::
+
+  /profile
+    /index.js // public interface
+    /ProfilePage.jsx
+    /ducks
+      /actions.js
+      /reducers.js
+      /selectors.js
+    /profile-photo
+      /index.js
+      /ProfilePhoto.jsx
+      /ducks
+        /actions.js
+        /reducers.js
+        /selectors.js
+    /education // Sparse sub-feature example
+      /index.js
+      /Education.jsx
+
+There may be other files in the /ducks directories as necessary.  e.g., sagas.js, operations.js, services.js, etc.
+
+**C. Features can have sub-features.**
+
+Sub-features should follow the same rules as any other feature.
+
+**D. A feature may import from outside its ancestry.**
+
+As described above in "Avoid Circular Dependencies", features should not import from their parent, grandparent, etc.  That said, they can import from siblings of their lineage.
+
+**E. A feature may not import from the children of its siblings.**
+
+If a feature needs something from the child of a sibling or peer, that child sub-feature may need to be pulled up a level (or more). Importing from some other feature's sub-feature is akin to violating the strict module boundaries described above, and could indicate that a refactoring is needed.
```

Comment:
How is one to tell which component in a different hierarchy is at the same peer level?  That is, if I need to depend on another component in another hierarchy, how can I tell whether that component is my peer or a child of my peer?

Reply:
This may not be very well-stated.  Put another way, you can't "reach into" another module.  That would violate its module boundary.  If a submodule needs to be shared by multiple modules, then it needs to be extracted from one so it's accessible by the other.  (Going to try to re-frame this)